pep795.md

```diff
@@ -0,0 +1,1938 @@
+# Deep Immutability for Efficient Sharing and Concurrency Safety
+
+This PEP proposes that Python objects can be made immutable after
+which attempts to mutate them results in an exception. We will
+refer to the act of making an object immutable as "freezing it".
+There is no way to make an immutable object mutable again.
+However, one can make a mutable copy of an immutable object.
+
+The motivation for immutability is four-fold: 
+
+1. Immutability is a very useful concept in programming that
+   simplifies programming. In particular immutable objects are
+   safe to share between concurrent threads without any
+   synchronisation in the program.
+2. Immutable objects can be shared directly by reference
+   across subinterpreters. This is a big performance boost
+   for subinterpreters as Python currently requires objects
+   to be transferred across subinterpreters by pickling.
+   This also simplifies programming with subinterpreters.
+3. Since immutable objects are safe from data races, it is
+   possible to avoid some of the overhead surrounding field access
+   that is needed for correct reference counting in free-threaded
+   Python.
+4. Memory for immutable objects can be safely managed without leaks
+   without cycle detection, even when there are cycles in the 
+   immutable objects. This takes pressure off normal cycle 
+   detection in both subinterpreters and free-threading, and
+   is key to sharing objects between subinterpreters without
+   needing substantial changes to cycle detection.
+
+In this PEP, we strive to create a semantics for immutable
+objects that is consistent across both subinterpreters and
+free-threaded Python. This is a difficult problem as
+subinterpreters rely on isolation plus a per-interpreter GIL for
+correctness, whereas free-threading has no GIL.
+
+
+# Semantics of Immutability
+
+This section describes the immutability semantics. All user-facing
+functions in this PEP are in the module `immutable`. The most
+important function is `freeze()` that makes an object immutable.
+Note that freezing is *in-place*; `freeze(obj)` freezes `obj` and
+for convenience also returns a reference to (now immutable) `obj`.
+
+```python
+class Person:
+    def __init__(self, name):
+        self.name = name
+
+monty = [Person("Eric"), Person("Graham"), Person("Terry")]
+freeze(monty)
+
+monty.append(Person("John")) # throws exception because monty is immutable
+monty.pop()                  # --''--
+
+monty[0].name = "John"       # throws exception because all person objects 
+                             # in monty are immutable too
+```
+
+Making an object immutable means that subsequent attempts
+to write to its fields will raise an exception. (The attempt
+to append or pop above will both result in `TypeError:
+object of type list is immutable`.) Furthermore, by default,
+immutability is **deep**. This means that an immutable
```

Comment:
This is missing an introduction, why immutability needs to be deep. Maybe the graph from the paper would be enough, where we show the problem of sharing with shallow immutability.

The rest of the section makes no sense, if people don't believe that immutability needs to be deep. For a lot of this, the obvious initial answer would be: "Just don't make immutability deep?"

Reply:
Great point. When I write about sharing objects directly across subinterpreters for the first time, I will point out that this needs deep immutability to work.